### PR TITLE
add explicit json annotation to dto properties

### PR DESCRIPTION
**Pull request by huangp**

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="http://app.review.ninja/zanata/zanata-api/pull/30" target="_blank">ReviewNinja

**Comment by seanf** [comment]
Looks good to me in principle, but it seems like it would be easy to break something.

If this were XML, we could generate the schema from the old version and from the new, to make sure they are the same.  As it is, we'll just have to hope we've got everything and/or that the compatibility tests are trustworthy.

Perhaps we could do something with http://wiki.fasterxml.com/JacksonJsonSchemaGeneration

**Comment by huangp** [comment]
I added the test to make use of the schema generation. However there are two cases we can't cover. It's in the test's comment

File: zanata-common-api/src/test/java/org/zanata/rest/dto/JsonSchemaTest.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                { TextFlowTarget.class },
+                { TranslationsResource.class },
+                { ContributionStatistics.class },
+                { LocaleStatistics.class },
+                // this one cuases stack overflow in schema generation
+                // { ContainerTranslationStatistics.class },
+                // This one has @XmlTransient on some of the getters and jackson
+                // will pick them up if JAXB annotation is used. There is no way
+                // we can be backward compatible (except turning jaxb off).
+                // { TranslationStatistics.class },
```
Comment (seanf) [inline zanata-common-api/src/test/java/org/zanata/rest/dto/JsonSchemaTest.java]: Isn't the point of `@XmlTransient` to tell JAXB to ignore the property?

What exactly do you mean when you say we can't be backwards compatible?  Do we have to abandon the whole idea?

**Comment by huangp** [comment]
It's not really a RESTEasy 2/3 problem. RESTEasy just provide a thin wrapper around Jackson as provider. Another way to solve "the server" problem is to configure it to continue ignoring JAXB annotation. But it won't solve the problem in our API (e.g. someone decided to use our API and choose JSON marshalling they will run into problem). I think we need to fix the DTO regardless. zanata-api compat is not really the problem here.

For now I will create sample DTO test for the one can't generate schema. The other one is a bug in our API (we should've always put JsonIgnore together with XmlTransient). It will not pass the test.

**Comment by huangp** [comment]
Tomas replied he is not using the stats API so we are "safe" to ignore those fields. Note the server resteasy 3 upgrade will no longer depend on this PR.

**Comment by seanf** [comment]
> Would it be possible to make use of zanata-api compat in this test? We don't want future enhancements to our DTOs (such as new properties) to break this test. Then again, I can't decide if we need to keep this test around long term.

@huangp any response to my earlier question?

**Comment by huangp** [comment]
I have answered it in my previous comment here https://github.com/zanata/zanata-api/pull/30#issuecomment-112604209 . 

It's testing JSON schema of the same version of API under different jackson marshalling policy (with jaxb annotation recognized or without). api-compat has nothing to do in this test. Future enhencements to DTO will only break this test if we make the same mistake again. e.g. forget to put @JsonIgnore where we have @XmlTransient. 

----- Original Message -----

**Comment by seanf** [comment]
>
>
> ----- Original Message -----
>
> > From: "Sean Flanigan" <notifications@github.com>
> > To: "zanata/zanata-api" <zanata-api@noreply.github.com>
> > Cc: "Patrick Huang" <pahuang@redhat.com>
> > Sent: Wednesday, August 5, 2015 4:00:49 PM
> > Subject: Re: [zanata-api] add explicit json annotation to dto properties
> > (#30)
>
> > > Would it be possible to make use of zanata-api compat in this test? We
> > > don't
> > > want future enhancements to our DTOs (such as new properties) to break
> this
> > > test. Then again, I can't decide if we need to keep this test around
> long
> > > term.
> >
> > @huangp any response to my earlier question?
>
>
On 5 August 2015 at 16:16, Patrick Huang <notifications@github.com> wrote:

> I have answered it in my previous comment here
> https://github.com/zanata/zanata-api/pull/30#issuecomment-112604209 .
>
> It's testing JSON schema of the same version of API under different
> jackson marshalling policy (with jaxb annotation recognized or without).


Sorry, it's been a while, and I've forgotten some details of the original
problem.  Do we really want to test that, specifically?

What causes JAXB annotations to be recognised or not?  Is it that RESTEasy
3 tells Jackson to use JAXB annotations for JSON by default, whereas
RESTEasy 2 does not?  So previously, Jackson was generating most of the
JSON properties by looking at the method names, instead of the JAXB
annotations?


> api-compat has nothing to do in this test. Future enhancements to DTO will
> only break this test

if we make the same mistake again. e.g. forget to put @JsonIgnore where we
> have @XmlTransient.
>

(Checking that we use @JsonIgnore wherever we have @XmlTransient might be a
good test.)

The thing is, we are not specifically concerned about making sure the JSON
schema is always the same under different Jackson policies.  But we are
concerned about backward compatibility of the API, which is exactly why
api-compat exists.

Isn't the point of the @JsonProperty annotations to make sure that Zanata
3.8 (under RESTEasy 3) uses the same JSON property names as Zanata 3.7
(under RESTEasy 2) did (which were accidentally derived from the method
names)?  We need tests for that, whether they live in zanata-api or in
zanata-war.


For instance, if we refactor the method name getReferences() in
PotEntryHeader to getSourceReferences(), keeping the same annotations, we
*don't* want the name of the JSON property to change (which it would have
under the RESTEasy 2 policy, by accident), and we should have a test
(somewhere) for that.

And I think the current schema test would actually fail in that case, which
we don't want, because it's testing the *updated* API under the RESTEasy 2
rules (JSON property names ultimately derived from method names), even
though the updated API is designed for RESTEasy 3 rules (JSON property
names ultimately derived from annotations).

Perhaps the schema test should compare the schema for zanata-api-compat
3.6(?) (using RESTEasy 2 rules) against latest zanata-api (or
zanata-api-compat 3.7(?)) using RESTEasy 3 rules.  Exactly which versions
to test would be something to be decided.  If we run into trouble because
of unrelated API changes between the API versions, we could check
PotEntryHeader and TextFlowTarget only, since it is they which are changing
in this PR.


-- 
Sean Flanigan

Principal Software Engineer
Globalisation Tools Engineering
Red Hat

**Comment by seanf** [comment]
:+1:

**Comment by definite** [comment]
As this PR is covered by unit testes and server arquillian tests. This one does not need QA.

**Pull request merged**